[R-package] Suppress 'warning: unknown pragma' for CRAN package

## Description of the problem

As of the changes in #3160 , builds of the CRAN version of the R package with `gcc` now raise "unknown pragma" warnings.

CRAN does not allow the use of `-Wunknown-pragmas` to suppress such warnings, so we remove them with `sed` when building the R package.

The "bug" here is that LightGBM's CI did not catch this. That should also be fixed.

## How to fix this

Add `pragma: warning` to the set of checks at https://github.com/microsoft/LightGBM/blob/5b0ab05c3d69eb9c0015dcadccaf8ad637b587c0/build-cran-package.sh#L53-L65.

Add some code to the CI script for Mac/Linux that tests for the presence of this warning and fails if it finds it. Similar to  https://github.com/microsoft/LightGBM/blob/5b0ab05c3d69eb9c0015dcadccaf8ad637b587c0/.ci/test_r_package.sh#L206-L216.

## How fixing this improves `{lightgbm}`

This warning does not affect the correctness of an installation, but showing it in logs may waste users' time by leading them to believe that something is wrong with their installation.

## Environment info

LightGBM version or commit hash: latest `master` (https://github.com/microsoft/LightGBM/commit/5b0ab05c3d69eb9c0015dcadccaf8ad637b587c0)

## Error message and / or logs

![image](https://user-images.githubusercontent.com/7608904/95003596-7294da80-05a6-11eb-8929-30b1afba79ce.png)

```text
clang++ -mmacosx-version-min=10.13 -std=gnu++11 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -I./include -DUSE_SOCKET -DLGB_R_BUILD  -I/usr/local/include  -Xclang -fopenmp -pthread -fPIC  -Wall -g -O2  -c treelearner/voting_parallel_tree_learner.cpp -o treelearner/voting_parallel_tree_learner.o
In file included from treelearner/voting_parallel_tree_learner.cpp:11:
In file included from treelearner/parallel_tree_learner.h:15:
treelearner/cuda_tree_learner.h:255:13: warning: unknown pragma ignored [-Wunknown-pragmas]
    #pragma warning(disable : 4702)
            ^
1 warning generated.
```

## Reproducible example(s)

On a Mac, using `gcc`:

```shell
sh build-cran-package.sh

R CMD INSTALL lightgbm_*.tar.gz
```

--- a/.ci/test_r_package.sh
+++ b/.ci/test_r_package.sh
@@ -226,7 +226,7 @@ if [[ $R_BUILD_TYPE == "cran" ]]; then
         | wc -l
     )
     if [[ $pragma_warning_present -ne 0 ]]; then
-        echo "Unknown pragma warning is present, pragmas should have been remove before build"
+        echo "Unknown pragma warning is present, pragmas should have been removed before build"
         exit -1
     fi
 fi

--- a/.ci/test_r_package.sh
+++ b/.ci/test_r_package.sh
@@ -219,7 +219,7 @@ fi
 
 # this check makes sure that no "warning: unknown pragma ignored" logs
 # reach the user leading them to believe that something went wrong
-if [[ $OS_NAME == "macos" || $OS_NAME == "linux" ]] && [[ $R_BUILD_TYPE == "cran" ]]; then
+if [[ $R_BUILD_TYPE == "cran" ]]; then
     pragma_warning_present=$(
         cat $BUILD_LOG_FILE \
         | grep -E "warning: unknown pragma ignored" \

--- a/.ci/test_r_package.sh
+++ b/.ci/test_r_package.sh
@@ -216,3 +216,17 @@ if [[ $OS_NAME == "macos" ]] && [[ $R_BUILD_TYPE == "cran" ]]; then
         exit -1
     fi
 fi
+
+# this check makes sure that no "warning: unknown pragma ignored" logs
+# reach the user leading them to believe that something went wrong
+if [[ $OS_NAME == "macos" || $OS_NAME == "linux" ]] && [[ $R_BUILD_TYPE == "cran" ]]; then
+    pragma_warning_present=$(
+        cat $BUILD_LOG_FILE \
+        | grep -E "warning: unknown pragma ignored" \
+        | wc -l
+    )
+    if [[ $pragma_warning_present -ne 0 ]]; then
+        echo "Unknown pragma warning is present, pragmas should have been remove before build"
+        exit -1
+    fi
+fi

--- a/build-cran-package.sh
+++ b/build-cran-package.sh
@@ -55,7 +55,7 @@ cd ${TEMP_R_DIR}
     # not allow you to use compiler flag '-Wno-unknown-pragmas' or
     # pragmas that suppress warnings.
     echo "Removing unknown pragmas in headers"
-    for file in $(find . -regex ".*\.\(h\|cpp\|hpp\)$"); do
+    for file in $(find . -regex '.*\.\(h\|cpp\|hpp\)$'); do
       sed \
         -i.bak \
         -e 's/^.*#pragma region.*$//' \
@@ -63,7 +63,7 @@ cd ${TEMP_R_DIR}
         -e 's/^.*#pragma warning.*$//' \
         "${file}"
     done
-    find . -regex ".*\.\(h\|cpp\|hpp\)\.bak$" -exec rm {} \;
+    find . -regex '.*\.\(h\|cpp\|hpp\)\.bak$' -exec rm {} \;
 
     # When building an R package with 'configure', it seems
     # you're guaranteed to get a shared library called

--- a/build-cran-package.sh
+++ b/build-cran-package.sh
@@ -50,19 +50,20 @@ cd ${TEMP_R_DIR}
     sed -i.bak -e "s/~~VERSION~~/${LGB_VERSION}/" DESCRIPTION
     sed -i.bak -e "s/~~DATE~~/${CURRENT_DATE}/" DESCRIPTION
 
-    # Remove 'region' and 'endregion' pragmas. This won't change
-    # the correctness of the code. CRAN does not allow you
-    # to use compiler flag '-Wno-unknown-pragmas' or
+    # Remove 'region', 'endregion', and 'warning' pragmas.
+    # This won't change the correctness of the code. CRAN does
+    # not allow you to use compiler flag '-Wno-unknown-pragmas' or
     # pragmas that suppress warnings.
     echo "Removing unknown pragmas in headers"
-    for file in src/include/LightGBM/*.h; do
+    for file in $(find . -name *.h); do
       sed \
         -i.bak \
         -e 's/^.*#pragma region.*$//' \
         -e 's/^.*#pragma endregion.*$//' \
+        -e 's/^.*#pragma warning.*$//' \
         "${file}"
     done
-    rm src/include/LightGBM/*.h.bak
+    find . -name *.h.bak -exec rm {} \;
 
     # When building an R package with 'configure', it seems
     # you're guaranteed to get a shared library called

--- a/build-cran-package.sh
+++ b/build-cran-package.sh
@@ -55,7 +55,7 @@ cd ${TEMP_R_DIR}
     # not allow you to use compiler flag '-Wno-unknown-pragmas' or
     # pragmas that suppress warnings.
     echo "Removing unknown pragmas in headers"
-    for file in $(find . -name *.h); do
+    for file in $(find . -regex ".*\.\(h\|cpp\|hpp\)$"); do
       sed \
         -i.bak \
         -e 's/^.*#pragma region.*$//' \
@@ -63,7 +63,7 @@ cd ${TEMP_R_DIR}
         -e 's/^.*#pragma warning.*$//' \
         "${file}"
     done
-    find . -name *.h.bak -exec rm {} \;
+    find . -regex ".*\.\(h\|cpp\|hpp\)\.bak$" -exec rm {} \;
 
     # When building an R package with 'configure', it seems
     # you're guaranteed to get a shared library called

--- a/build-cran-package.sh
+++ b/build-cran-package.sh
@@ -55,7 +55,7 @@ cd ${TEMP_R_DIR}
     # not allow you to use compiler flag '-Wno-unknown-pragmas' or
     # pragmas that suppress warnings.
     echo "Removing unknown pragmas in headers"
-    for file in $(find . -regex '.*\.\(h\|cpp\|hpp\)$'); do
+    for file in $(find . -name '*.h' -o -name '*.hpp' -o -name '*.cpp'); do
       sed \
         -i.bak \
         -e 's/^.*#pragma region.*$//' \
@@ -63,7 +63,7 @@ cd ${TEMP_R_DIR}
         -e 's/^.*#pragma warning.*$//' \
         "${file}"
     done
-    find . -regex '.*\.\(h\|cpp\|hpp\)\.bak$' -exec rm {} \;
+    find . -name '*.h.bak' -o -name '*.hpp.bak' -o -name '*.cpp.bak' -exec rm {} \;
 
     # When building an R package with 'configure', it seems
     # you're guaranteed to get a shared library called

